Behaviour default views

jaxwilko opened the pull request:
This PR adds default views for the following backend controller behaviors:

- FormController
- ImportExportController
- ListController
- ReorderController

`Backend\Classes\ControllerBehavior` will now automatically append their `views` folder to the controller's view paths allowing them to provide fallbacks for any views required by the behavior.

The `create:controller` command will now no longer generate the views by default unless `--stubs` is also passed and the `--sidebar` flag is replaced with a `--layout=(standard|sidebar|fancy)` option to choose the form layout to use. 

Also:
- Added `appendViewPath()` and `prependViewPath()` to the `System\Traits\ViewMaker`. `addViewPath` is renamed to `prependViewPath()` and is for paths that have higher priority than the existing paths while `appendViewPath()` is for paths that should have lower priority than the existing paths (i.e. fallbacks). 
- Backend controllers will now automatically set their navigation context in the form of `Author.Plugin` as the author, `$pluginName` as the main menu code, and `$controllerName` as the side menu code. This means that you can remove calls to `BackendMenu::setContext()` and constructor overrides in your controllers if they follow that convention.
- Support for passing `new: true` as a parameter in the request body to `onSave()` calls that will return a redirect to the `create` action
- `formMakePartial(string $partial, array $params = [])` to the `FormController` behavior that will render a partial through the controller's `makePartial` using the following priority list of contextual names: `form_$context_$partial`, `form_$partial`, `$partial`).
- Support for `abort(403)` to return the access denied view in the backend
- Improved handling of `abort(404)` in the backend
- Improved styling of disabled fields in the fancy form layout
- Improved styling of file generated / updated status message in scaffolding commands

To Review:
- [x] Core controllers in Backend, CMS, & System modules
- [x] Flexible form buttons (https://github.com/wintercms/winter/pull/1352)
- [x] Winter.Blog (https://github.com/wintercms/wn-blog-plugin/pull/72)
- [x] Winter.Forum (https://github.com/wintercms/wn-forum-plugin/pull/17)
- [x] Winter.Location (https://github.com/wintercms/wn-location-plugin/pull/7)
- [x] Winter.Notify (https://github.com/wintercms/wn-notify-plugin/pull/3)
- [x] Winter.Redirect (https://github.com/wintercms/wn-redirect-plugin/pull/23)
- [x] Winter.Sitemap (https://github.com/wintercms/wn-sitemap-plugin/pull/22)
- [x] Winter.SSO (https://github.com/wintercms/wn-sso-plugin/pull/15)
- [x] Winter.Test (https://github.com/wintercms/wn-test-plugin/pull/22)
- [x] Winter.Translate (https://github.com/wintercms/wn-translate-plugin/pull/95)
- [x] Winter.User (https://github.com/wintercms/wn-user-plugin/pull/57)
- [x] Add support for this to the scaffolding commands

jaxwilko committed:
Added support for pre/post appending view paths

jaxwilko committed:
Added default view files for forms and lists behaviours

jaxwilko committed:
Added --stubs option to create controller command

jaxwilko committed:
Added code styling fixes

jaxwilko committed:
Added fix for hardcoded link

mjauvin commented:
I've been wanting to do this for ages...

Great work!

mjauvin commented on modules/backend/behaviors/listcontroller/_list_toolbar.php:
```diff
@@ -0,0 +1,26 @@
+
+<div data-control="toolbar">
+    <a
+        href="<?= url()->current() . '/create' ?>"
```
```suggestion
        href="<?= $this->actionUrl('create') ?>"
```

mjauvin commented on modules/backend/behaviors/formcontroller/update.php:
```diff
@@ -0,0 +1,55 @@
+                </button>
+                <span class="btn-text">
+                    or <a href="<?= Backend::url($formConfig->defaultRedirect) ?>"><?= e(trans('backend::lang.form.cancel')); ?></a>
+                </span>
```
```suggestion
                </span>
                <?= $this->fireViewEvent('backend.form.extendUpdateView.buttons') ?>
```

LukeTowers replied:
@mjauvin I'd prefer if this was handled the same way that the RelationController handles list toolbar buttons with specific partial names that can be used / a config item in the config_form.yaml.

mjauvin commented on modules/backend/behaviors/listcontroller/_list_toolbar.php:
```diff
@@ -0,0 +1,26 @@
+        data-request-success="$(this).prop('disabled', 'disabled')"
+        data-stripe-load-indicator>
+        <?= e(trans('backend::lang.list.delete_selected')); ?>
+    </button>
```
```suggestion
    </button>
    <?= $this->fireViewEvent('backend.list.extendListToolbar') ?>
```

LukeTowers replied:
@mjauvin I'd prefer if the list toolbar buttons were configurable more in the way that the relationcontroller handles buttons instead of a generic view event.

jaxwilko replied:
You can also just add `_list_toolbar.php` to your controller's folder and it'll be overwritten

LukeTowers commented on modules/backend/console/CreateController.php:
```diff
@@ -69,11 +68,16 @@ protected function prepareVars(): array
-        $this->stubs["scaffold/controller/{$layout}/create.stub"] = 'controllers/{{lower_name}}/create.php';
-        $this->stubs["scaffold/controller/{$layout}/update.stub"] = 'controllers/{{lower_name}}/update.php';
-        $this->stubs["scaffold/controller/{$layout}/preview.stub"] = 'controllers/{{lower_name}}/preview.php';
+            $layout = $this->option('sidebar') ? 'sidebar' : 'standard';
```
We'll need to add support for this as a property on the controller class ($formLayout perhaps?) and then have the default views look at that property before rendering and have the property set in the scaffold. 

Options would be "standard" (can be not set), "sidebar" (the my account page with secondaryTabs rendered on the right), or "fancy" (the CMS / static pages layout).

LukeTowers committed:
Update modules/backend/behaviors/listcontroller/_list_toolbar.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

jaxwilko committed:
Added support for default reorder views

jaxwilko committed:
Merge branch 'wip/behaviour-default-views' of github.com:wintercms/winter into wip/behaviour-default-views

jaxwilko commented:
@LukeTowers @mjauvin any thoughts on my changes to `ViewMaker`?

mjauvin commented:
> @LukeTowers @mjauvin any thoughts on my changes to `ViewMaker`?

I like it @jaxwilko, as long as it doesn't introduce BC issues, which it shouldn't.

jaxwilko committed:
Added support for import export behaviour defaults

jaxwilko committed:
Added default value for bodyClass

jaxwilko committed:
Added self generating breadcrumb partial

jaxwilko committed:
Added default partials for behaviours

jaxwilko committed:
Added self generating breadcumb for controller stubs

jaxwilko committed:
Added import export lang keys

jaxwilko committed:
Added fix for code styling

jaxwilko committed:
Added fixes for reorder partials

LukeTowers committed:
Move controllerbehavior default views to /views

LukeTowers committed:
Merge branch 'develop' into wip/behaviour-default-views

LukeTowers commented on modules/backend/behaviors/reordercontroller/views/reorder.php:
```diff
@@ -0,0 +1,5 @@
+<?php Block::put('breadcrumb') ?>
+    <?= $this->makeLayoutPartial('breadcrumb') ?>
+<?php Block::endPut() ?>
+
```
```suggestion
<!-- Reorder Controller Widget -->
```

LukeTowers committed:
Update modules/backend/behaviors/reordercontroller/views/reorder.php

LukeTowers committed:
Switch to new prependViewPath() method

LukeTowers committed:
Automatically set BackendMenu context from controller path

Credit to @sephyld. Replaces #1311. Closes #1269.

LukeTowers committed:
Tweaks to the default views based on migrating Winter.Blog

mjauvin commented:
This is such a great improvement, will make renaming a controller much less tedious!

mjauvin commented:
Running this on all controllers for my latest project in dev, so far so good!

mjauvin commented:
@LukeTowers anything preventing this to be merged?

LukeTowers commented:
@mjauvin I need to finish going through the rest of the Winter plugins and testing it out to make sure we're not missing any pieces. Already had some valuable changes just from implementing it on the Blog plugin so I want to check the remaining ones. Will be working on it again tomorrow during office hours.

LukeTowers committed:
Improve breadcrumb generation logic to take into account settings contexts

LukeTowers committed:
Update the changemonitor status after the AJAX requests complete

LukeTowers committed:
Add top padding to form buttons container & remove buttons from the sidebar preview view

LukeTowers committed:
Remove unnecessary backend controller views

LukeTowers committed:
Add localized "or" string

LukeTowers committed:
Add data-browser-validate to all form saving buttons

mjauvin committed:
add missing delete button to sidebar layout update view

mjauvin commented:
@jaxwilko I added the missing "delete (trash)" icon to the default update view for the sidebar layout here:

https://github.com/wintercms/winter/commit/5de7bd7a87ca201de8e4dccaa92cdef8094f49df

mjauvin commented:
@jaxwilko should I commit the fix below?

```diff
diff --git a/modules/backend/layouts/_breadcrumb.php b/modules/backend/layouts/_breadcrumb.php
index c9f575539..cc56ddfda 100644
--- a/modules/backend/layouts/_breadcrumb.php
+++ b/modules/backend/layouts/_breadcrumb.php
@@ -12,7 +12,7 @@
 <?php if ($menu): ?>
 <ul>
     <li><a href="<?= $menu->url ?>"><?= e(trans($menu->label)) ?></a></li>
-    <?php if ($sideMenu): ?>
+    <?php if ($sideMenu && $sideMenu->url !== $menu->url): ?>
         <li><a href="<?= $sideMenu->url ?>"><?= e(trans($sideMenu->label)) ?></a></li>
     <?php elseif ($settingsItem): ?>
         <li><a href="<?= $settingsItem->url ?>"><?= e(trans($settingsItem->label)) ?></a></li>
```

See here for an explanation:
https://github.com/wintercms/wn-user-plugin/pull/57#issuecomment-2726943868

jaxwilko commented:
> @jaxwilko should I commit the fix below?
> 
> ```diff
> diff --git a/modules/backend/layouts/_breadcrumb.php b/modules/backend/layouts/_breadcrumb.php
> index c9f575539..cc56ddfda 100644
> --- a/modules/backend/layouts/_breadcrumb.php
> +++ b/modules/backend/layouts/_breadcrumb.php
> @@ -12,7 +12,7 @@
>  <?php if ($menu): ?>
>  <ul>
>      <li><a href="<?= $menu->url ?>"><?= e(trans($menu->label)) ?></a></li>
> -    <?php if ($sideMenu): ?>
> +    <?php if ($sideMenu && $sideMenu->url !== $menu->url): ?>
>          <li><a href="<?= $sideMenu->url ?>"><?= e(trans($sideMenu->label)) ?></a></li>
>      <?php elseif ($settingsItem): ?>
>          <li><a href="<?= $settingsItem->url ?>"><?= e(trans($settingsItem->label)) ?></a></li>
> ```
> 
> See here for an explanation:
> https://github.com/wintercms/wn-user-plugin/pull/57#issuecomment-2726943868

Yeah go ahead and commit it, everything needs final testing anyway which i'm aiming to do next week :)

mjauvin committed:
no need for two links with the same url

mjauvin committed:
Merge branch 'develop' into wip/behaviour-default-views

mjauvin commented:
@LukeTowers any plans to merge this in the near future ?

mjauvin commented:
@LukeTowers I updated all plugins and left `Winter.Test` for you to handle... :stuck_out_tongue:

mjauvin committed:
do not show "delete selected" button if the list does not show the records checkboxes

mjauvin committed:
add cancel button in default fancy toolbar

mjauvin committed:
remove extra space after closing parenthesis

LukeTowers commented on modules/backend/behaviors/listcontroller/views/_list_toolbar.php:
```diff
@@ -0,0 +1,28 @@
+        <?= e(trans('backend::lang.form.create_title', ['name' => trans(\Winter\Storm\Support\Str::before($listConfig->title, '_plural'))])); ?>
+    </a>
+
+  <?php if (isset($listConfig->showCheckboxes) && $listConfig->showCheckboxes != false): ?>
+    <button
+        class="btn btn-danger wn-icon-trash-o"
+        disabled="disabled"
+        onclick="$(this).data('request-data', { checked: $('.control-list').listWidget('getChecked') })"
+        data-request="onDelete"
+        data-request-confirm="<?= e(trans('backend::lang.list.delete_selected_confirm')); ?>"
+        data-trigger-action="enable"
+        data-trigger=".control-list input[type=checkbox]"
+        data-trigger-condition="checked"
+        data-request-success="$(this).prop('disabled', 'disabled')"
+        data-stripe-load-indicator>
+        <?= e(trans('backend::lang.list.delete_selected')); ?>
+    </button>
+  <?php endif ?>
```
```suggestion
    <?php if (isset($listConfig->showCheckboxes) && $listConfig->showCheckboxes != false): ?>
        <button
            class="btn btn-danger wn-icon-trash-o"
            disabled="disabled"
            onclick="$(this).data('request-data', { checked: $('.control-list').listWidget('getChecked') })"
            data-request="onDelete"
            data-request-confirm="<?= e(trans('backend::lang.list.delete_selected_confirm')); ?>"
            data-trigger-action="enable"
            data-trigger=".control-list input[type=checkbox]"
            data-trigger-condition="checked"
            data-request-success="$(this).prop('disabled', 'disabled')"
            data-stripe-load-indicator
        >
            <?= e(trans('backend::lang.list.delete_selected')); ?>
        </button>
    <?php endif ?>
```

LukeTowers committed:
Update modules/backend/behaviors/listcontroller/views/_list_toolbar.php

mjauvin committed:
Merge branch 'develop' into wip/behaviour-default-views

LukeTowers committed:
Merge branch 'develop' into wip/behaviour-default-views

LukeTowers committed:
Merge branch 'develop' into wip/behaviour-default-views

mjauvin committed:
Extract create/update form buttons into their own partials (#1352)

Co-authored-by: Luke Towers <git@luketowers.ca>

There is now a single partial for the formcontroller create/update buttons.

This partial can be overridden in your controllers using the partial names below:

create: _form_create_toolbar.php
update: _form_update_toolbar.php
both: _form_toolbar.php

Also adds Create and new button/action to FormController.

LukeTowers committed:
Use toolbar partial

LukeTowers committed:
Update update.stub

LukeTowers committed:
Update create.stub

LukeTowers committed:
Update update.stub

mjauvin committed:
add reorder button on default list toolbar

mjauvin committed:
fix icon class

mjauvin committed:
only add create button if FormController is used

mjauvin committed:
use more appropriate reorder icon

mjauvin committed:
use arrows-up-down icon for reorder action

mjauvin committed:
Merge branch 'develop' into wip/behaviour-default-views

mjauvin committed:
Merge branch 'develop' into wip/behaviour-default-views

mjauvin commented:
@jaxwilko @LukeTowers I'd like to add the possibility to not show the Trash (delete) button in the `_toolbar.php` formController partial if a property of the controller is present (e.g. $hideTrashButton). I've always found that button can be mistaken for the relation manager delete button (i.e. one selects an item in the relation manager and instead of clicking "delete relation item" one clicks the form "trash/delete" button.

What do you guys think ?

```diff
diff --git a/modules/backend/behaviors/formcontroller/partials/_toolbar.php b/modules/backend/behaviors/formcontroller/partials/_toolbar.php
index d4929319f..f07addfa4 100644
--- a/modules/backend/behaviors/formcontroller/partials/_toolbar.php
+++ b/modules/backend/behaviors/formcontroller/partials/_toolbar.php
@@ -65,15 +65,17 @@ class="btn btn-default"
         >
             <?= e(trans('backend::lang.form.save_and_close')); ?>
         </button>
-        <button
-            type="button"
-            data-request="onDelete"
-            data-load-indicator="<?= e(trans('backend::lang.form.deleting_name', ['name' => trans($modelName)])); ?>"
-            data-request-before-update="$el.trigger('unchange.oc.changeMonitor')"
-            data-request-confirm="<?= e(trans('backend::lang.form.confirm_delete')); ?>"
-            class="wn-icon-trash-o btn-icon danger pull-right"
-        >
-        </button>
+        <?php if (!$this->hideTrashButton): ?>
+            <button
+                type="button"
+                data-request="onDelete"
+                data-load-indicator="<?= e(trans('backend::lang.form.deleting_name', ['name' => trans($modelName)])); ?>"
+                data-request-before-update="$el.trigger('unchange.oc.changeMonitor')"
+                data-request-confirm="<?= e(trans('backend::lang.form.confirm_delete')); ?>"
+                class="wn-icon-trash-o btn-icon danger pull-right"
+            >
+            </button>
+        <?php endif ?>
         <span class="btn-text">
             <?= e(trans('backend::lang.form.or')) ?> <a href="<?= Backend::url($formConfig->defaultRedirect) ?>"><?= e(trans('backend::lang.form.cancel')); ?></a>    
         </span>
```

LukeTowers commented:
@mjauvin if we make it configurable then it should probably be through the form config and it should also disable the Ajax handler. Could we change the delete button to make it more obvious what it does instead?

mjauvin commented:
> @mjauvin if we make it configurable then it should probably be through the form config and it should also disable the Ajax handler. Could we change the delete button to make it more obvious what it does instead?

Can we do both ?

```php
<?php if (!isset($formConfig->hideTrashButton)): ?>
    <button
        type="button"
        data-request="onDelete"
        data-load-indicator="<?= e(trans('backend::lang.form.deleting_name', ['name' => trans($modelName)])); ?>"
        data-request-before-update="$el.trigger('unchange.oc.changeMonitor')"
        data-request-confirm="<?= e(trans('backend::lang.form.confirm_delete')); ?>"
        class="wn-icon-trash-o btn-icon danger pull-right"
    >
    </button>
<?php endif ?>
```

LukeTowers committed:
Merge branch 'develop' into wip/behaviour-default-views

LukeTowers commented:
@mjauvin why do you want to disable the button / the handler separately from each other? We might also want to take inspiration from the RelationController's approach to managing the available buttons (see `toolbarButtons` on https://wintercms.com/docs/v1.2/docs/backend/relations) if we go down this route. Either way, I'm thinking we should make that a separate PR so that we can review it without blocking this one.

LukeTowers committed:
Add buttons for import export behavior to list default toolbar when behavior is present

LukeTowers committed:
Add support for abort(403) in the backend

LukeTowers committed:
Improve abort(404) handling in the backend

jaxwilko commented on modules/backend/behaviors/FormController.php:
```diff
@@ -882,4 +888,20 @@ public static function extendFormFields($callback)
+     */
+    public function formMakePartial(string $partial, array $params = []): string
+    {
+        $contents = $this->controller->makePartial('form_' . $this->context . '_' . $partial, $params + $this->vars, false);
```
```suggestion
        $contents = $this->controller->makePartial('form_' . $this->context . '_' . $partial, [...$params, ...$this->vars], false);
```
?

jaxwilko commented on modules/backend/behaviors/FormController.php:
```diff
@@ -882,4 +888,20 @@ public static function extendFormFields($callback)
+    {
+        $contents = $this->controller->makePartial('form_' . $this->context . '_' . $partial, $params + $this->vars, false);
+        if (!$contents) {
+            $contents = $this->controller->makePartial('form_' . $partial, $params + $this->vars, false);
```
```suggestion
            $contents = $this->controller->makePartial('form_' . $partial, array_merge([], $params, $this->vars), false);
```
?

LukeTowers commented:
@mjauvin thanks for all your work on this, it's made it much easier to review. I'll made it through half of the related PRs today, should hopefully be able to finish it up during office hours next week. Many thanks to @jaxwilko for the improved `abort()` handling in the backend as well ❤️

LukeTowers committed:
Add icons to the default form buttons

LukeTowers committed:
Merge branch 'develop' into wip/behaviour-default-views

LukeTowers committed:
Add missing buttons and icons to the form fancy toolbars

LukeTowers committed:
Fix disabled styling for outside tab fields in the fancy form layout

Also recompiled the other less assets since the dependency update to the less compiler.

LukeTowers committed:
Improve handling of layouts in the controller scaffolding command
